invokeai/app/services/session_queue/session_queue_sqlite.py

```diff
@@ -0,0 +1,845 @@
+                    item_id TEXT NOT NULL PRIMARY KEY, -- the unique identifier of this queue item
+                    order_id INTEGER NOT NULL, -- used for ordering, cursor pagination
+                    batch_id TEXT NOT NULL, -- identifier of the batch this queue item belongs to
+                    queue_id TEXT NOT NULL, -- identifier of the queue this queue item belongs to
+                    session_id TEXT NOT NULL UNIQUE, -- duplicated data from the session column, for ease of access
+                    field_values TEXT, -- NULL if no values are associated with this queue item
+                    session TEXT NOT NULL, -- the session to be executed
+                    status TEXT NOT NULL DEFAULT 'pending', -- the status of the queue item, one of 'pending', 'in_progress', 'complete', 'error', 'canceled'
+                    priority INTEGER NOT NULL DEFAULT 0, -- the priority, higher is more important
+                    error TEXT, -- any errors associated with this queue item
+                    created_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')),
+                    updated_at DATETIME NOT NULL DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')), -- updated via trigger
+                    started_at DATETIME, -- updated via trigger
+                    completed_at DATETIME -- updated via trigger, completed items are cleaned up on application startup
+                    -- Ideally this is a FK, but graph_executions uses INSERT OR REPLACE, and REPLACE triggers the ON DELETE CASCADE...
+                    -- FOREIGN KEY (session_id) REFERENCES graph_executions (id) ON DELETE CASCADE
+                );
+                """
+            )
+
+            self.__cursor.execute(
+                """--sql
+                CREATE UNIQUE INDEX IF NOT EXISTS idx_session_queue_item_id ON session_queue(item_id);
+                """
+            )
+
+            self.__cursor.execute(
+                """--sql
+                CREATE UNIQUE INDEX IF NOT EXISTS idx_session_queue_order_id ON session_queue(order_id);
+                """
+            )
+
+            self.__cursor.execute(
+                """--sql
+                CREATE UNIQUE INDEX IF NOT EXISTS idx_session_queue_session_id ON session_queue(session_id);
+                """
+            )
+
+            self.__cursor.execute(
+                """--sql
+                CREATE INDEX IF NOT EXISTS idx_session_queue_batch_id ON session_queue(batch_id);
+                """
+            )
+
+            self.__cursor.execute(
+                """--sql
+                CREATE INDEX IF NOT EXISTS idx_session_queue_created_priority ON session_queue(priority);
+                """
+            )
+
+            self.__cursor.execute(
+                """--sql
+                CREATE INDEX IF NOT EXISTS idx_session_queue_created_status ON session_queue(status);
+                """
+            )
+
+            self.__cursor.execute(
```

Comment:
Curious why there are triggers here - are the statuses changed anywhere outside of this service? Wondering if it would also work to just update these columns within the class functions that change status

Reply:
No, not changed from outside the service.

There are 4 different queries that set the status, some of which set it for N queue items, and one of which can set to any status. That's a lot of places to miss a case, so I opted to centralize the logic here